content/blog/client_node_and_cncf_mentorship.md

```diff
@@ -0,0 +1,208 @@
+
+## Repo structure
+
+I first started writing Rust wrappers on Raw client - a low level API for querying/updating individual key-value pairs. The NodeJS client has a dependency on Rust client because that's one of the benefits using Neon - to reuse the code in Rust client. 
+
+The `client-node` repo structure looks like this:
+
+```rust
+├── Cargo.lock
+├── Cargo.toml
+├── index.node
+├── package.json
+├── src
+    └── lib.rs
+		... 
+├── tikv_client
+		└── index.js
+```
+
+The `[lib.rs](http://lib.rs)` implements the main logic that exposes the client functionalities. In each client method, the program parses all function arguments and pass them to native rust-client function. Then returns the result. Here is the an example:
+
+```rust
+pub fn put(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+	  let client = cx
+	      .this()
+	      .downcast_or_throw::<JsBox<RawClient>, _>(&mut cx)?;
+	  let key = cx.argument::<JsString>(0)?.value(&mut cx);
+	  let value = cx.argument::<JsString>(1)?.value(&mut cx);
+	  let cf = cx.argument::<JsString>(2)?.value(&mut cx);
+	  let callback = cx.argument::<JsFunction>(3)?.root(&mut cx);
+	
+	  let inner = client.inner.with_cf(cf.try_into().unwrap());
+	  let channel = cx.channel();
+	
+	  RUNTIME.spawn(async move {
+	      let result = inner.put(key, value).await;
+	      send_result(channel, callback, result);
+	  });
+	
+	  Ok(cx.undefined())	
+}
+```
+
+The first variable "client" saves a pointer to the Client object living in the heap. It is a pointer to `JsBox`. The variables "key", "value", "cf" are string typed parameters required to call `put` method on rust client. The "callback" is the last parameter, also a function, for sending callbacks. The "inner" variable is the rust-client struct, which is initialized with the configuration. The "channel" variable provides a channel for sending events back to JavaScript. Then it uses a async runtime to drive spawn a new thread, and asynchronously calls the `put` method from rust-client, saves the result to "result" variable, and send it back to JavaScript channel. 
+
+There are a lot to unpack from the above example. Before I talk about the details, let me show what the corresponding function in JavaScript looks like
+
+```jsx
+put(key: string, value: string, cf: string) {
+  return put.call(this.boxed, key, value, cf);
+}
+```
+
+The above JavaScript code a simple wrapper for `put` function, specifies all the arguments needed and passes them to the native Rust client `put_async` function. It does not have any logic of doing the actual work. Below is a usage example when the library is loaded to Nodejs:
+
+```jsx
+(async () => {
+  const client = await new tikv.RawClient("127.0.0.1:2379");
+  await client.put("k1", "v1", "default");
+})();
+```
```

Comment:
I would suggest moving this section up before the rust code implementation details. And then it could attract the reader by telling them with an example of calling a rust library function using nodejs code (maybe we can also add the rust client example to show what the nodejs code is reflecting to), and after that, we explain how we achieve that.